### ISSUE TITLE
mat_add, mat_vec the matrix A representation

### ISSUE THREAD

**Issue by nishalad95** [uwhpsc-2016/homework2#12]
In the `linalg.c` file that we need to write our functions in, the comments mention representing Matrices A, B and out as long arrays. I understand that this is so we can access each element of the matrix in a contiguous way going through row by row of the matrix. But what i'm confused at is do we need to change the type from `double*` to `long`? As it says we need to represent A, B and out as arrays of length `M x N`, or can we simply make a for loop starting from 0 to `M x N` adding up each corresponding element and storing it in the matrix out?

**Comment by cswiercz** [comment]
The signature of `mat_add` is 

```c
// include.h
void mat_add(double* out, double* A, double* B, int M, int N);
```

The elements of the matrices `A` and `B` are of type `double`. The function requests the dimensions `M` and `N` so that we know the length of the rows (`N`) and how many rows there are (`M`). You need these data if you want to add the two matrices together and store the result in `out`.

I don't know why you want to change the type of the matrices to `long` but perhaps the above clarifies things?

**Comment by afarahatoutset** [comment]
Probably confused "long" array with type long.

**Comment by cswiercz** [comment]
No worries. Also, reopening so other students can see this discussion.

**Comment by nishalad95** [comment]
Ah thank you! I think that's made more sense now!

**Comment by cswiercz** [comment]
Awesome. Glad to help.

**Comment by atom-moyer** [comment]
To clarify, when you refer to a "long matrix" in the write up, you mean a 1D matrix that contains the same amount of values of a 2D MxN matrix, which I suspect because we are only passing 1 pointer for all of the matrixes instead of passing an array of pointers which defines a 2D matrix.

**Comment by cswiercz** [comment]
> To clarify, when you refer to a "long matrix" in the write up, you mean a 1D matrix that contains the same amount of values of a 2D MxN matrix

Correct. Very often in C when you want to store an `M x N` matrix you store it as an array containing `MN` elements where the first `N` elements represent row 0, the next `N` elements represent row 1, etc.

> which I suspect because we are only passing 1 pointer for all of the matrixes instead of passing an array of pointers which defines a 2D matrix.

Yes. You may have found online some people using an array of pointers to represent a matrix so that you can use the familiar indexing notation

`A[i][j]`

The downside to this format is that
- if you want to heap-allocate `A` you need to heap-allocate `M` arrays of length `N`, which is `M` times as much memory to have to keep track of,
- contiguity of memory access is broken at the end of each row. That is, a routine like `mat_add` can never be completely contiguous if you choose to store your matrices in this way.

In the land of high-performance computing you want to avoid these kinds of weaknesses. However, in the land of having to spend human hours writing code there is a balance between choosing a data format / design which is easy to read an understand and squeezing the last bit of performance out of your code.
